src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java

```diff
@@ -0,0 +1,53 @@
+package se.liu.ida.hefquin.engine.queryplan.physical.impl;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.jena.sparql.core.Var;
+
+import se.liu.ida.hefquin.engine.queryplan.ExpectedVariables;
+import se.liu.ida.hefquin.engine.queryplan.executable.impl.ops.BinaryExecutableOp;
+import se.liu.ida.hefquin.engine.queryplan.executable.impl.ops.ExecOpBinaryUnion;
+import se.liu.ida.hefquin.engine.queryplan.logical.BinaryLogicalOp;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpUnion;
+import se.liu.ida.hefquin.engine.queryplan.physical.BinaryPhysicalOpForLogicalOp;
+
+public class PhysicalOpBinaryUnion implements BinaryPhysicalOpForLogicalOp{
+	
+	protected final LogicalOpUnion lop;
+
+	protected PhysicalOpBinaryUnion(LogicalOpUnion lop) {
+		assert lop != null;
+		this.lop = lop;
+	}
+
+	@Override
+	public ExpectedVariables getExpectedVariables(ExpectedVariables... inputVars) {
+		// perhaps this should be offered as the default implementation in the interface.
+		
+		assert inputVars.length == 2;
+
+		final Set<Var> certainVars = new HashSet<>( inputVars[0].getCertainVariables());
+		final Set<Var> possibleVars = new HashSet<>( inputVars[0].getPossibleVariables() );
+
+		certainVars.addAll(inputVars[1].getCertainVariables());
+		possibleVars.addAll(inputVars[1].getPossibleVariables());
+		possibleVars.removeAll(certainVars);
+
+		return new ExpectedVariables() {
+			@Override public Set<Var> getCertainVariables() { return certainVars;}

```

Comment:
That's incorrect. The set of certain variables of a union operator is the _intersection_ of the sets of certain variables of the two inputs (rather than the union of these two sets, as in your current implementation).

Reply:
So, my question now is regarding the semantics of the union operator. At first I thought that it would be similar to SPARQL union, where you can have a query like this, taken from the SPARQL specification:
```
SELECT ?x ?y
WHERE  { { ?book dc10:title ?x } UNION { ?book dc11:title  ?y } }
```
Where the certain variables are the union of the input variables. So, please, can you show me what I'm getting wrong?

Reply:
Yes, our union operator works exactly like union in SPARQL. I assume you have read the definition of "certain variables" and "possible variables" that I had added in the comments in the ExpectedVariables interface. Certain variables of an operator are all the variables that can be expected to be bound in _every_ solution mapping produced by that operator. In your example, that would be only ?book but not ?x and not ?y. In general, it is the intersection of the certain variables of the two input operators.

Reply:
Right, I get it now, thanks!

Reply:
This looks good now. Thanks for fixing!